File: src/ServerScriptService/ServerSource/Server/CatService/Components/Others/InteractionHandler.lua
local InteractionHandler = {}

-- External dependencies
local CatProfileData

-- Internal state
local InteractionHandler = {
	PendingInteractions = {},
	InteractionCooldowns = {}
}

function InteractionHandler:HandleInteraction(player, catId, interactionType, interactionData)
	-- Check cooldown
	if InteractionHandler:IsOnCooldown(player, catId, interactionType) then
		return {success = false, message = "Interaction on cooldown"}
	end
	
	-- Get cat data
	local CatService = script.Parent.Parent.Parent
	local catData = CatService.GetComponent:GetCat(catId)
	if not catData then
		return {success = false, message = "Cat not found"}
	end
	
	-- Get interaction effects
	local interactionEffects = CatProfileData.GetInteractionEffects(interactionType)
	
	-- Calculate success chance based on relationship and mood
	local successChance = InteractionHandler:CalculateSuccessChance(player, catId, interactionType, catData)
	
	-- Determine outcome
	local success = math.random() <= successChance
	local result = {
		success = success,
		interactionType = interactionType,
		effects = {}
	}
	
	if success then
		-- Apply successful interaction effects
		result.effects = InteractionHandler:ApplySuccessfulInteraction(player, catId, interactionType, catData, interactionEffects)
		result.message = "Interaction successful!"
	else
		-- Apply failed interaction effects
		result.effects = InteractionHandler:ApplyFailedInteraction(player, catId, interactionType, catData)
		result.message = "Interaction failed - cat was not interested"
	end
	
	-- Set cooldown
	InteractionHandler:SetCooldown(player, catId, interactionType)
	
	-- Log interaction
	InteractionHandler:LogInteraction(player, catId, interactionType, success, result.effects)
	
	print(string.format("Interaction: %s with cat %s - Success: %s", interactionType, catId, success))
	
	return result
end

function InteractionHandler:CalculateSuccessChance(player, catId, interactionType, catData)
	local baseChance = CatProfileData.GetInteractionEffects(interactionType).successChance or 0.5
	
	-- Relationship modifier
	local relationship = InteractionHandler:GetPlayerRelationship(player, catId)
	local relationshipModifier = 0.5 + (relationship.trustLevel or 0.5)
	
	-- Mood modifier
	local moodEffects = CatProfileData.GetMoodEffects(catData.moodState.currentMood)
	local moodModifier = moodEffects.interactionChance or 1.0
	
	-- Personality modifier
	local personality = catData.profile.personality
	local personalityModifier = 1.0
	
	if interactionType == "Pet" then
		personalityModifier = personality.friendliness
	elseif interactionType == "Play" then
		personalityModifier = personality.playfulness
	elseif interactionType == "Feed" then
		personalityModifier = 1.0 -- All cats like food
	end
	
	-- Final success chance
	local successChance = baseChance * relationshipModifier * moodModifier * personalityModifier
	
	-- Clamp between 0.1 and 0.95
	return math.clamp(successChance, 0.1, 0.95)
end

function InteractionHandler:ApplySuccessfulInteraction(player, catId, interactionType, catData, interactionEffects)
	local effects = {}
	
	-- Update relationship
	local relationshipChange = interactionEffects.relationshipChange or 0.1
	InteractionHandler:UpdateRelationship(player, catId, relationshipChange)
	effects.relationshipChange = relationshipChange
	
	-- Update mood
	if interactionEffects.moodEffect then
		local CatService = script.Parent.Parent.Parent
		CatService.SetComponent:UpdateCatMood(catId, interactionEffects.moodEffect, 0.7)
		effects.moodEffect = interactionEffects.moodEffect
	end
	
	-- Update physical state
	if interactionEffects.hungerReduction then
		catData.physicalState.hunger = math.max(0, catData.physicalState.hunger - interactionEffects.hungerReduction)
		effects.hungerReduction = interactionEffects.hungerReduction
	end
	
	if interactionEffects.energyCost then
		catData.physicalState.energy = math.max(0, catData.physicalState.energy - interactionEffects.energyCost)
		effects.energyCost = interactionEffects.energyCost
	end
	
	-- Update cat's current action
	catData.behaviorState.currentAction = interactionType
	catData.behaviorState.isInteracting = true
	
	-- Set interaction timeout
	task.delay(3, function()
		catData.behaviorState.isInteracting = false
		if catData.behaviorState.currentAction == interactionType then
			catData.behaviorState.currentAction = "Idle"
		end
	end)
	
	return effects
end

function InteractionHandler:ApplyFailedInteraction(player, catId, interactionType, catData)
	local effects = {}
	
	-- Small negative relationship impact
	local relationshipChange = -0.05
	InteractionHandler:UpdateRelationship(player, catId, relationshipChange)
	effects.relationshipChange = relationshipChange
	
	-- Possible mood change to annoyed
	if math.random() < 0.3 then
		local CatService = script.Parent.Parent.Parent
		CatService.SetComponent:UpdateCatMood(catId, "Annoyed", 0.4)
		effects.moodEffect = "Annoyed"
	end
	
	return effects
end

function InteractionHandler:GetPlayerRelationship(player, catId)
	-- Get the parent CatService to access relationship manager
	local CatService = script.Parent.Parent.Parent
	return CatService.GetComponent:GetPlayerRelationship(player, catId) or {
		trustLevel = 0.5,
		interactionHistory = {},
		lastInteraction = 0
	}
end

function InteractionHandler:UpdateRelationship(player, catId, change)
	-- Get the parent CatService to update relationship
	local CatService = script.Parent.Parent.Parent
	CatService.SetComponent:UpdatePlayerRelationship(player, catId, change)
end

function InteractionHandler:IsOnCooldown(player, catId, interactionType)
	local key = player.UserId .. "_" .. catId .. "_" .. interactionType
	local cooldownEnd = InteractionHandler.InteractionCooldowns[key]
	
	if cooldownEnd and os.time() < cooldownEnd then
		return true
	end
	
	return false
end

function InteractionHandler:SetCooldown(player, catId, interactionType)
	local key = player.UserId .. "_" .. catId .. "_" .. interactionType
	local cooldownDuration = 2 -- seconds
	
	InteractionHandler.InteractionCooldowns[key] = os.time() + cooldownDuration
	
	-- Clean up old cooldowns
	for k, endTime in pairs(InteractionHandler.InteractionCooldowns) do
		if os.time() > endTime then
			InteractionHandler.InteractionCooldowns[k] = nil
		end
	end
end

function InteractionHandler:LogInteraction(player, catId, interactionType, success, effects)
	local relationship = InteractionHandler:GetPlayerRelationship(player, catId)
	
	-- Add to interaction history
	table.insert(relationship.interactionHistory, {
		type = interactionType,
		timestamp = os.time(),
		outcome = success and "positive" or "negative",
		effects = effects
	})
	
	-- Keep only last 50 interactions
	if #relationship.interactionHistory > 50 then
		table.remove(relationship.interactionHistory, 1)
	end
	
	relationship.lastInteraction = os.time()
	
	-- Update favorite activities
	if success then
		InteractionHandler:UpdateFavoriteActivities(relationship, interactionType)
	end
end

function InteractionHandler:UpdateFavoriteActivities(relationship, interactionType)
	relationship.favoriteActivities = relationship.favoriteActivities or {}
	
	-- Count frequency of successful interactions
	local activityCounts = {}
	for _, interaction in ipairs(relationship.interactionHistory) do
		if interaction.outcome == "positive" then
			activityCounts[interaction.type] = (activityCounts[interaction.type] or 0) + 1
		end
	end
	
	-- Get top 3 favorite activities
	local favorites = {}
	for activity, count in pairs(activityCounts) do
		table.insert(favorites, {activity = activity, count = count})
	end
	
	table.sort(favorites, function(a, b) return a.count > b.count end)
	
	relationship.favoriteActivities = {}
	for i = 1, math.min(3, #favorites) do
		table.insert(relationship.favoriteActivities, favorites[i].activity)
	end
end

-- Component initialization
function InteractionHandler.Init()
	-- Load dependencies
	local ReplicatedStorage = game:GetService("ReplicatedStorage")
	local sharedDatas = ReplicatedStorage:WaitForChild("SharedSource").Datas
	CatProfileData = require(sharedDatas.CatProfileData)
	
	print("InteractionHandler component initialized")
end

function InteractionHandler.Start()
	print("InteractionHandler component started")
end

return InteractionHandler

File: src/server/interaction-handler.ts
import { CatManager } from "./cat-manager";
import { RelationshipManager } from "./relationship-manager";
import { CatProfileData } from "shared/cat-profile-data";
import { CatAI } from "./cat-ai";
import { Players } from "@rbxts/services";
import { INTERACTION_TYPES } from "shared/config/behavior-config";
import { CatData, InteractionEffect } from "shared/cat-types";

export class InteractionHandler {
    private static interactionCooldowns = new Map<string, number>();

    public static HandleInteraction(
        player: Player,
        catId: string,
        interactionType: string,
        interactionData?: unknown,
    ): { success: boolean; message: string; interactionType?: string } {
        if (this.IsOnCooldown(player, catId, interactionType)) {
            return { success: false, message: "Interaction on cooldown" };
        }

        const catData = CatManager.GetCat(catId);
        if (!catData) {
            return { success: false, message: "Cat not found" };
        }

        const effects = INTERACTION_TYPES[interactionType];

        // Special case: Release if already held by this player
        if (interactionType === "Hold" && catData.behaviorState.heldByPlayerId === player.UserId) {
            this.ApplyRelease(catId, catData);
            return { success: true, message: "Released cat", interactionType: "Hold" };
        }

        if (catData.behaviorState.heldByPlayerId !== undefined) {
            return { success: false, message: "Cat is already being held" };
        }

        if (!effects) {
            return { success: false, message: "Invalid interaction type" };
        }

        const successChance = this.CalculateSuccessChance(player, catId, interactionType, catData);
        const success = math.random() <= successChance;

        const result = {
            success,
            interactionType,
            message: success ? "Interaction successful!" : "Interaction failed - cat was not interested",
        };

        if (success) {
            this.ApplySuccessfulInteraction(player, catId, interactionType, catData, effects);
        } else {
            this.ApplyFailedInteraction(player, catId, catData);
        }

        this.SetCooldown(player, catId, interactionType);
        return result;
    }

    private static CalculateSuccessChance(player: Player, catId: string, interactionType: string, catData: CatData): number {
        const effects = INTERACTION_TYPES[interactionType];
        const baseChance = effects?.successChance || 0.5;

        const relationship = RelationshipManager.GetRelationship(player, catId);
        const relationshipModifier = 0.5 + relationship.trustLevel;

        const moodModifier = CatProfileData.GetMoodEffects(catData.moodState.currentMood).interactionChance || 1.0;

        let personalityModifier = 1.0;
        const personality = catData.profile.personality;

        if (interactionType === "Pet") {
            personalityModifier = personality.friendliness;
        } else if (interactionType === "Play") {
            personalityModifier = personality.playfulness;
        } else if (interactionType === "Hold") {
            personalityModifier = (personality.friendliness + (1 - personality.shyness)) / 2;
        }

        const successChance = baseChance * relationshipModifier * moodModifier * personalityModifier;
        return math.clamp(successChance, 0.1, 0.95);
    }

    private static ApplySuccessfulInteraction(
        player: Player,
        catId: string,
        interactionType: string,
        catData: CatData,
        effects: InteractionEffect,
    ) {
        const relChange = effects.relationshipChange || 0.1;
        const relationship = RelationshipManager.UpdateRelationship(player, catId, relChange);

        // Sync relationship to catData so client can access it
        if (!catData.socialState.playerRelationships.has(player.UserId)) {
            catData.socialState.playerRelationships.set(player.UserId, relationship);
        } else {
            // Update existing relationship
            const existingRel = catData.socialState.playerRelationships.get(player.UserId)!;
            existingRel.trustLevel = relationship.trustLevel;
            existingRel.relationshipScore = relationship.relationshipScore;
            existingRel.relationshipTier = relationship.relationshipTier;
            existingRel.lastInteraction = relationship.lastInteraction;
        }

        if (effects.moodEffect) {
            CatManager.UpdateCatMood(catId, effects.moodEffect, 0.7);
        }

        if (effects.hungerReduction !== undefined) {
            CatManager.UpdateCatPhysical(catId, { hunger: -effects.hungerReduction });
        }

        if (effects.energyCost !== undefined) {
            CatManager.UpdateCatPhysical(catId, { energy: -effects.energyCost });
        }

        if (interactionType === "Hold") {
            catData.behaviorState.heldByPlayerId = player.UserId;
            catData.behaviorState.isMoving = false;
        }

        // Special reaction for Pet interaction: cat looks at player and purrs, then stays near them
        if (interactionType === "Pet") {
            const char = player.Character;
            const hrp = char?.FindFirstChild("HumanoidRootPart") as Part;
            if (hrp) {
                // Store that this player recently petted the cat
                const aiData = CatAI.GetAIData(catId);
                if (aiData) {
                    aiData.memory.set("LastPettedBy", player.UserId);
                    aiData.memory.set("LastPettedTime", os.time());
                    aiData.memory.set("StayNearPlayerUntil", os.time() + 15); // Stay near for 15 seconds
                }

                // Initial purr reaction (3 seconds)
                catData.behaviorState.currentAction = "Purr";
                catData.behaviorState.targetPosition = hrp.Position;
                catData.behaviorState.isMoving = false;
                catData.behaviorState.actionData = { reactingToPlayerId: player.UserId };
                
                // After purring, transition to staying near player
                task.delay(3, () => {
                    const updatedCatData = CatManager.GetCat(catId);
                    const updatedAiData = CatAI.GetAIData(catId);
                    if (updatedCatData && updatedAiData) {
                        const stayUntil = updatedAiData.memory.get("StayNearPlayerUntil") as number | undefined;
                        const pettedPlayerId = updatedAiData.memory.get("LastPettedBy") as number | undefined;
                        
                        // If still within the stay-near window, transition to Follow/LookAt
                        if (stayUntil && os.time() < stayUntil && pettedPlayerId) {
                            const pettedPlayer = Players.GetPlayerByUserId(pettedPlayerId);
                            if (pettedPlayer?.Character?.PrimaryPart) {
                                updatedCatData.behaviorState.currentAction = "Follow";
                                updatedCatData.behaviorState.targetPosition = pettedPlayer.Character.PrimaryPart.Position;
                                updatedAiData.memory.set("SocialTarget", pettedPlayerId);
                            }
                        } else {
                            // Time window expired, return to normal behavior
                            updatedCatData.behaviorState.currentAction = "Idle";
                            updatedCatData.behaviorState.actionData = undefined;
                        }
                    }
                });

                // Clear the stay-near state after the full duration
                task.delay(15, () => {
                    const updatedCatData = CatManager.GetCat(catId);
                    const updatedAiData = CatAI.GetAIData(catId);
                    if (updatedCatData && updatedAiData) {
                        const stayUntil = updatedAiData.memory.get("StayNearPlayerUntil") as number | undefined;
                        // Only clear if this is still the same petting session
                        if (stayUntil && os.time() >= stayUntil) {
                            updatedAiData.memory.delete("LastPettedBy");
                            updatedAiData.memory.delete("LastPettedTime");
                            updatedAiData.memory.delete("StayNearPlayerUntil");
                            
                            // If cat is still following/looking at this player, transition to idle
                            if (updatedCatData.behaviorState.currentAction === "Follow" || 
                                updatedCatData.behaviorState.currentAction === "LookAt") {
                                const currentTarget = updatedAiData.memory.get("SocialTarget") as number | undefined;
                                if (currentTarget === player.UserId) {
                                    updatedCatData.behaviorState.currentAction = "Idle";
                                    updatedAiData.memory.delete("SocialTarget");
                                }
                            }
                        }
                    }
                });
            }
        }

        RelationshipManager.AddInteractionToHistory(player, catId, {
            type: interactionType,
            timestamp: os.time(),
            outcome: "positive",
            effects: {},
        });
    }

    private static ApplyFailedInteraction(player: Player, catId: string, catData: CatData) {
        const relationship = RelationshipManager.UpdateRelationship(player, catId, -0.05);

        // Sync relationship to catData so client can access it
        if (!catData.socialState.playerRelationships.has(player.UserId)) {
            catData.socialState.playerRelationships.set(player.UserId, relationship);
        } else {
            // Update existing relationship
            const existingRel = catData.socialState.playerRelationships.get(player.UserId)!;
            existingRel.trustLevel = relationship.trustLevel;
            existingRel.relationshipScore = relationship.relationshipScore;
            existingRel.relationshipTier = relationship.relationshipTier;
            existingRel.lastInteraction = relationship.lastInteraction;
        }

        if (math.random() < 0.3) {
            CatManager.UpdateCatMood(catId, "Annoyed", 0.4);
        }

        RelationshipManager.AddInteractionToHistory(player, catId, {
            type: "failed_interaction",
            timestamp: os.time(),
            outcome: "negative",
            effects: {},
        });
    }

    private static ApplyRelease(catId: string, catData: CatData) {
        catData.behaviorState.heldByPlayerId = undefined;
    }

    private static IsOnCooldown(player: Player, catId: string, interactionType: string): boolean {
        const key = `${player.UserId}_${catId}_${interactionType}`;
        const cooldownEnd = this.interactionCooldowns.get(key);
        return cooldownEnd !== undefined && os.time() < cooldownEnd;
    }

    private static SetCooldown(player: Player, catId: string, interactionType: string) {
        const key = `${player.UserId}_${catId}_${interactionType}`;
        this.interactionCooldowns.set(key, os.time() + 2);
    }

    // For testing: Clear cooldown for a specific interaction
    public static ClearCooldown(player: Player, catId: string, interactionType: string): void {
        const key = `${player.UserId}_${catId}_${interactionType}`;
        this.interactionCooldowns.delete(key);
    }
}
